### PR TITLE
Refactor Search with video and audio

### PR DESCRIPTION
**Pull request by Regis18**
Search video and audio

### CODE REVIEW

File: src/main/java/com/jala/model/search/SearchAudio.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import net.bramp.ffmpeg.FFprobe;
+import net.bramp.ffmpeg.probe.FFmpegProbeResult;
+import org.apache.log4j.Logger;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * SearchAudio
```
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchAudio.java]: Complete documentation please.

File: src/main/java/com/jala/model/search/SearchAudio.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                }
+                if (resultSearchAudio) {
+                    asset = AssetFactory.getAsset(preview.get(i),
+                            ffprobeResult.getStreams().get(0).codec_long_name,
+                            ffprobeResult.getStreams().get(0).channel_layout,
+                            Integer.toString(ffprobeResult.getStreams().get(0).sample_rate),
+                            String.valueOf(ffprobeResult.getStreams().get(0).duration));
+                    result.add(asset);
+                }
+
```
Comment (raulchoque-eng) [inline src/main/java/com/jala/model/search/SearchAudio.java]: Clear this line.

File: src/main/java/com/jala/model/search/SearchBasic.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+ * @version 0.0.3
+ * @author Regis Humana
+ */
+public abstract class SearchBasic implements ISearchable {
+    /** It creates to follow up the instruction of the class*/
+    private static Logger log = Logs.getInstance().getLog();
+    /** asset list*/
+    private List<Asset> result;
+
+	/**
```
Comment (raulchoque-eng) [inline src/main/java/com/jala/model/search/SearchBasic.java:50]: Justify this comment.

File: src/main/java/com/jala/model/search/SearchCommon.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+ * @version 0.0.3
+ * @author Regis Humana
+ */
+public class SearchCommon extends SearchBasic {
+    /** It creates to follow up the instruction of the class*/
+    private Logger log = Logs.getInstance().getLog();
+    private CriteriaSearch criteria;
+    private List<Asset> result;
+
+	/**
```
Comment (raulchoque-eng) [inline src/main/java/com/jala/model/search/SearchCommon.java:42]: Justify this comment.

File: src/main/java/com/jala/model/search/SearchAudio.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        List<Asset> preview = super.search();
+        for (int i = 0; i < preview.size(); i++) {
+            FFprobe ffprobe;
+            FFmpegProbeResult ffprobeResult;
+            try {
+                ffprobe = new FFprobe(FFPROBE_PATH);
+                ffprobeResult = ffprobe.probe(preview.get(i).getPath());
+                boolean resultSearchAudio = true;
+                Asset asset;
+                if (resultSearchAudio && !(ffprobeResult.getStreams().get(0).codec_long_name).contains(criteriaSearchAudio.getAudioCodec())) {
```
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchAudio.java]: `get(0)` the same parameter is used in other line please create a constant for this parameter

File: src/main/java/com/jala/model/search/SearchCommon.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+import java.util.List;
+
+/**
+ * SearchCommon
+ *
+ * @version 0.0.3
+ * @author Regis Humana
+ */
+public class SearchCommon extends SearchBasic {
+    /** It creates to follow up the instruction of the class*/
```
Comment (raulchoque-eng) [inline src/main/java/com/jala/model/search/SearchCommon.java]: Add a line above this line..

File: src/main/java/com/jala/model/search/SearchAudio.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+                    if (!criteriaSearchAudio.getAudioCodec().equals("")) {
```
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchAudio.java]: create constant to empty value "" or use the methos `empty()`

File: src/main/java/com/jala/model/search/SearchAudio.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+                    }
+                }
+                if (!criteriaSearchAudio.getAudioSampleRate().equals("")) {
+                    double sampleRate = Double.parseDouble(criteriaSearchAudio.getAudioSampleRate());
+                    if (resultSearchAudio && !(sampleRate >= ffprobeResult.getStreams().get(0).sample_rate)) {
+                        resultSearchAudio = false;
+                    }
+                }
+                if (!criteriaSearchAudio.getDuration().equals("")) {
+                    double durationCriteria = Double.parseDouble(criteriaSearchAudio.getDuration()) * 60;
```
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchAudio.java]: use constant for value 60 please

File: src/main/java/com/jala/model/search/SearchFile.java [hunk 9]
```diff
@@ -12,134 +12,53 @@
...
     /** It creates to follow up the instruction of the class*/
     private Logger log = Logs.getInstance().getLog();
 
     /**
-     * This method return a IAsset list by attributes of criteria.
-     * @param criteria to do the search with the path is required and other attributes are optional.
-     * @return a list of the files that content the folder set in the path of criteria.
+     * Searching for commons files.
+     *
+     * @param criteria
```
Comment (raulchoque-eng) [inline src/main/java/com/jala/model/search/SearchFile.java:37]: Complete this comment.

File: src/main/java/com/jala/model/search/SearchBasic.java [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+
+/**
+ * SearchBasic
+ *
+ * @version 0.0.3
+ * @author Regis Humana
+ */
+public abstract class SearchBasic implements ISearchable {
+    /** It creates to follow up the instruction of the class*/
+    private static Logger log = Logs.getInstance().getLog();
```
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchBasic.java:45]: add a empty line between attributes

File: src/main/java/com/jala/model/search/SearchBasic.java (hunk 3 above)
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchBasic.java:50]: format you documentation please

File: src/main/java/com/jala/model/search/SearchBasic.java [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+                    result.add(asset);
+                }
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Get the file creation date.
+     * @param filePath the path of the file.
```
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchBasic.java:85]: add a empty line between documentation and param

File: src/main/java/com/jala/model/search/SearchBasic.java [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+     * @return the creation date of the file.
+     */
+    public static String fileCreationDate(String filePath) {
```
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchBasic.java]: the name of method start with verb please rename this method

File: src/main/java/com/jala/model/search/SearchBasic.java [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+        File file = new File(filePath);
+        BasicFileAttributes attrs;
```
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchBasic.java]: use a name specific please in attrs

File: src/main/java/com/jala/model/search/SearchVideo.java [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+	}
+
+	/**
+	 * Receive all the file that found in Search common, and deliver a list with the filters that was made for the
+	 * customer, in special video
+	 * @return List<Asset>
+	 */
+	@Override
+	public List<Asset> search() {
+		log.info("The search for Video is beginning");
```
Comment (raulchoque-eng) [inline src/main/java/com/jala/model/search/SearchVideo.java:68]: Justify only four spaces.

File: src/main/java/com/jala/model/search/SearchBasic.java [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+        String formatted;
+        try {
+            attrs = Files.readAttributes(file.toPath(), BasicFileAttributes.class);
+            FileTime time = attrs.creationTime();
+            String pattern = "yyyy-MM-dd";
+            SimpleDateFormat simpleDateFormat = new SimpleDateFormat(pattern);
+            formatted = simpleDateFormat.format( new java.util.Date( time.toMillis() ) );
```
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchBasic.java]: format in the "( )"

File: src/main/java/com/jala/model/search/SearchBasic.java [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+        }
+        return formatted;
+    }
+
+    /**
+     * Get the file last modified date.
+     * @param filePath the path of the file.
+     * @return the date of file the las modified date.
+     */
+    public static String fileLastModifiedDate(String filePath) {
```
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchBasic.java]: rename the method please this should start with a verb

File: src/main/java/com/jala/model/search/SearchCommon.java [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+import com.jala.utils.Logs;
+import org.apache.commons.io.FilenameUtils;
+import org.apache.log4j.Logger;
+
+import java.sql.Date;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * SearchCommon.
```
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchCommon.java:26]: add documentation for classes

File: src/main/java/com/jala/model/search/SearchCommon.java [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+ *
+ * @version 0.0.3
+ * @author Regis Humana
+ */
+public class SearchCommon extends SearchBasic {
+    /** It creates to follow up the instruction of the class. */
+    private Logger log = Logs.getInstance().getLog();
+    private CriteriaSearch criteria;
```
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchCommon.java:37]: add documentation for attributes

File: src/main/java/com/jala/model/search/asset/Asset.java [hunk 19]
```diff
@@ -201,4 +188,17 @@ public String getLastDate() {
     public void setLastDate(String lastDate) {
         this.lastDate = lastDate;
     }
+    /**
```
Comment (raulchoque-eng) [inline src/main/java/com/jala/model/search/asset/Asset.java:213]: Add a line above this line.

File: src/main/java/com/jala/model/search/SearchCommon.java (hunk 18 above)
Comment (AKatrin) [reply src/main/java/com/jala/model/search/SearchCommon.java:37]: add a empty line between attributes

File: src/main/java/com/jala/model/search/SearchCommon.java [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+        result = new ArrayList<>();
+        for (int i = 0; i < preview.size(); i++) {
+            if ((criteria.getHidden() == TernaryBooleanEnum.OnlyTrue) && !(preview.get(i).isHidden())) {
+                continue;
+            }
+            if ((criteria.getHidden() == TernaryBooleanEnum.OnlyFalse) && (preview.get(i).isHidden())) {
+                continue;
+            }
+            if ((!criteria.getCreationDateFrom().isEmpty()) && !criteria.getCreationDateTo().isEmpty()) {
+                if ((Date.valueOf(fileCreationDate(preview.get(i).getPath())).before(Date.valueOf(criteria.getCreationDateFrom())))
```
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchCommon.java]: cut line please it is very long

File: src/main/java/com/jala/model/search/SearchFile.java [hunk 21]
```diff
@@ -12,134 +12,53 @@
...
 
+import java.util.List;
+
 /**
- * SearchFile class.
- * @author Areliez Vargas
- * @version 0.0.1
+ * SearchFile.
+ *
+ * @author Regis Humana.
```
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchFile.java]: is need to change the author???

File: src/main/java/com/jala/model/search/SearchFile.java [hunk 22]
```diff
@@ -12,134 +12,53 @@
...
+        SearchCommon searchCommon = new SearchCommon(criteria);
+        return searchCommon.search();
     }
 
     /**
-     * Search files recursively.
-     * @param currentFile the starting file.
-     * @param result list of files.
+     * Searching for special files, like video and audio.
+     * @param criteria
```
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchFile.java]: add documentation for params

File: src/main/java/com/jala/model/search/SearchVideo.java [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+import org.apache.log4j.Logger;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * SearchVideo.
```
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchVideo.java:31]: Add documentation for description of class please

File: src/main/java/com/jala/model/search/SearchVideo.java [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+ * @author Regis Humana
+ * @version 0.0.3
+ */
+public class SearchVideo extends SearchCommon {
+	/**
+	 * Asset list.
+	 */
+	private List<Asset> result;
+
+	/** Criteria multimedia. */
```
Comment (AKatrin) [inline src/main/java/com/jala/model/search/SearchVideo.java:42]: what should be used to documentation of attributes 
`/**
* Asset list.
*/`
or
/** Criteria multimedia. */

**Review by AKatrin — CHANGES_REQUESTED** [review]

File: src/main/java/com/jala/model/search/asset/AssetAudio.java [hunk 25]
```diff
@@ -12,17 +12,16 @@
...
 /**
- * AssetVideo class.
+ * AssetAudio class.
  *
  * @author Melvi Caballero.
  * @version 0.0.1
  */
-public class AssetAudio {
+public class AssetAudio extends Asset {
     /**
```
Comment (raulchoque-eng) [inline src/main/java/com/jala/model/search/asset/AssetAudio.java]: Add a line above.

File: src/main/java/com/jala/model/search/asset/AssetFactory.java [hunk 26]
```diff
@@ -12,182 +12,98 @@
 
 package com.jala.model.search.asset;
 
-import com.jala.model.criteria.CriteriaSearch;
-import org.apache.commons.io.FilenameUtils;
```
Comment (raulchoque-eng) [inline src/main/java/com/jala/model/search/asset/AssetFactory.java]: Clear this line.

File: src/main/java/com/jala/model/search/asset/AssetVideo.java [hunk 27]
```diff
@@ -23,8 +23,8 @@
     private String audioCodec;
     private String frameRate;
     private String aspectRatio;
-    private String dimentionWidth;
-    private String dimentionHeight;
+    private String dimensionWidth;
+    private String dimensionHeight;
     private String audioSampleRate;
     private String duration;
```
Comment (raulchoque-eng) [inline src/main/java/com/jala/model/search/asset/AssetVideo.java:43]: Clear this line.

File: src/main/java/com/jala/model/search/asset/AssetVideo.java [hunk 28]
```diff
@@ -101,34 +101,34 @@ public void setAspectRatio(String aspectRatio) {
 
     /**
      *
```
Comment (raulchoque-eng) [inline src/main/java/com/jala/model/search/asset/AssetVideo.java]: Clear this line.

File: src/main/java/com/jala/utils/AssetFactory.java [hunk 29]
```diff
@@ -0,0 +1,112 @@
...
+ * This software is the confidential and proprietary information of
+ * Jala Foundation, ("Confidential Information").  You shall not
+ * disclose such Confidential Information and shall use it only in
+ * accordance with the terms of the license agreement you entered into
+ * with Jala Foundation.
+ */
+
+package com.jala.utils;
+
+
```
Comment (raulchoque-eng) [inline src/main/java/com/jala/utils/AssetFactory.java]: Clear this line.

File: src/test/java/com/jala/model/search/SearchAudioTest.java [hunk 30]
```diff
@@ -0,0 +1,133 @@
...
+ * This software is the confidential and proprietary information of
+ * Jala Foundation, ("Confidential Information").  You shall not
+ * disclose such Confidential Information and shall use it only in
+ * accordance with the terms of the license agreement you entered into
+ * with Jala Foundation.
+ */
+
+package com.jala.model.search;
+
+
```
Comment (raulchoque-eng) [inline src/test/java/com/jala/model/search/SearchAudioTest.java]: Clear this line.

File: src/test/java/com/jala/model/search/SearchCommonTest.java [hunk 31]
```diff
@@ -0,0 +1,92 @@
...
+ * accordance with the terms of the license agreement you entered into
+ * with Jala Foundation.
+ */
+
+package com.jala.model.search;
+
+import com.jala.model.criteria.CriteriaSearch;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
```
Comment (raulchoque-eng) [inline src/test/java/com/jala/model/search/SearchCommonTest.java]: Put the complete import.

File: src/test/java/com/jala/model/search/SearchVideoTest.java [hunk 32]
```diff
@@ -0,0 +1,149 @@
...
+		criteria = new CriteriaSearchMultimedia("C:\\Users\\Admin\\Desktop\\PROG 02\\Videos");
+	}
+
+	/**
+	 * Extension : mp4
+	 * Channel: stereo
+	 */
+	@Test
+	public void search_returnMP4_WhenSendPathMultimedia() {
+		criteria.setExtension("mp4");
```
Comment (raulchoque-eng) [inline src/test/java/com/jala/model/search/SearchVideoTest.java:45]: Justify only four spaces.

File: src/main/java/com/jala/model/search/SearchCommon.java (hunk 6 above)
Comment (AKatrin) [reply src/main/java/com/jala/model/search/SearchCommon.java]: please add dot in documentation

File: src/main/java/com/jala/model/criteria/CriteriaConverterVideo.java [hunk 33]
```diff
@@ -12,7 +12,6 @@
 
 package com.jala.model.criteria;
```
Comment (raulchoque-eng) [inline src/main/java/com/jala/model/criteria/CriteriaConverterVideo.java:14]: Clear this line.

Comment (raulchoque-eng) [reply src/main/java/com/jala/model/criteria/CriteriaConverterVideo.java:14]: Then clear the line 162

File: src/main/java/com/jala/model/criteria/CriteriaSearch.java [hunk 34]
```diff
@@ -14,6 +14,7 @@
 
 import com.jala.model.search.TernaryBooleanEnum;
 
+
```
Comment (raulchoque-eng) [inline src/main/java/com/jala/model/criteria/CriteriaSearch.java]: Clear this line.

File: src/main/java/com/jala/model/search/SearchBasic.java (hunk 3 above)
Comment (raulchoque-eng) [reply src/main/java/com/jala/model/search/SearchBasic.java:50]: Justify this comment.

File: src/main/java/com/jala/model/search/SearchCommon.java (hunk 4 above)
Comment (raulchoque-eng) [reply src/main/java/com/jala/model/search/SearchCommon.java:42]: Justify this comment.

File: src/main/java/com/jala/model/search/SearchCommon.java [hunk 35]
```diff
@@ -0,0 +1,112 @@
...
+	/**
+	 * This constructor receive the criteria, for using in the search.
+     * @param criteria
+	 */
+    public SearchCommon(CriteriaSearch criteria) {
+        log.info("Searching on " + criteria.getPath());
+        this.criteria = criteria;
+    }
+
+	/**
```
Comment (raulchoque-eng) [inline src/main/java/com/jala/model/search/SearchCommon.java:51]: Justify this comment.

File: src/main/java/com/jala/model/search/asset/AssetAudio.java [hunk 36]
```diff
@@ -12,27 +12,30 @@
 
 package com.jala.model.search.asset;
 /**
```
Comment (raulchoque-eng) [inline src/main/java/com/jala/model/search/asset/AssetAudio.java:15]: Add a line above.

File: src/main/java/com/jala/model/search/asset/AssetVideo.java (hunk 27 above)
Comment (raulchoque-eng) [reply src/main/java/com/jala/model/search/asset/AssetVideo.java:43]: Clear this line.

**Pull request merged**